### DISCUSSION TITLE
Extract data from files at compile time

### DISCUSSION THREAD

**Discussion by tavoyne** [webpack/webpack#15243] [Q&A] [answered]
Hello there,

I'm building a basic blog. In my source folder, I have a `posts` folder in which I store posts in the markdown (`.md`) format. Each post contains some metadata (e.g. `creationDate`, `title`, etc.) at the top, which can be extracted and parsed easily with a few lines of Node.js code. I want Webpack to create, at compile time, a mapping of the post paths to their respective metadata object and to make that mapping available at runtime. The reason for this is that, at runtime, I want to only import the 10 most recent posts without having to load everything upfront just to figure out what those are.

I was able to achieve the expected result using `webpack.DefinePlugin`. Code looks like that:
```javascript
new webpack.DefinePlugin({
  POSTS: webpack.DefinePlugin.runtimeValue(
    () => {
      // Finds all posts in the posts folder. Parses their metadata.
      // Returns an object like:
      // {
      //   'post-1': { creationDate: '2022/01/24', title: 'Post 1' },
      //   'post-2': { creationDate: '2022/01/23', title: 'Post 2' }
      // }
    },
    { contextDependencies: [path.resolve("src", "posts")],
  })
});
```

What's great here is that it works well with `webpack-dev-server`, as the mapping is recomputed every time a post changes, thanks to `contextDependencies`.

BUT, the approach of setting this data to a global variable doesn't seem very straightforward to me, and because the problem I'm trying to solve is certainly pretty common, I'm wondering if there's a more direct approach (or an existing plugin) that I'm not seeing here. I was thinking of making it a plugin.

Would love to hear your opinion.

**Comment by alexander-akait** [comment, marked as answer]
You can try to implement this using https://webpack.js.org/api/loaders/#thisimportmodule, i.e. execute module and extract necessary things (but this may require code rework)

**Reply by tavoyne** [reply]
Yes indeed, thanks a lot!
